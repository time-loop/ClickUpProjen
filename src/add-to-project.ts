import { typescript, YamlFile } from 'projen';

export module addToProjectWorkflow {
  export const RENOVATE_GITHUB_USERNAME = 'cu-infra-svc-git';

  export const DEFAULT_RENOVATE_PR_LABEL = 'renovate';
  export const GITHUB_PROJECT_NUMBER = '3';

  const defaultWorkflow = {
    name: 'add-to-project',
    on: {
      pull_request: {
        types: ['opened', 'labeled'],
      },
    },
    // Prevent renovate running multiple times in parallel due to several subsequent trigger events occurring
    concurrency: {
      group: "${{ github.workflow }}-${{ github.event_name == 'workflow_dispatch' && github.event_name || '' }}",
    },
    jobs: {
      'add-to-project': {
        'runs-on': 'ubuntu-latest',
        permissions: {
          contents: 'read',
        },
        steps: [
          {
            name: 'Checkout',
            uses: 'actions/checkout@v3',
          },
          {
            name: 'Add issue to project',
            uses: 'actions/add-to-project@v0.4.1',
            with: {
              // github project url
              'project-url': `https://github.com/orgs/time-loop/projects/${GITHUB_PROJECT_NUMBER}`,
              'github-token': 'x-access-token:${{ secrets.PROJEN_GITHUB_TOKEN }}',
              labeled: DEFAULT_RENOVATE_PR_LABEL,
            },
          },
        ],
      },
    },
  };

  export function addAddToProjectWorkflowYml(project: typescript.TypeScriptProject, override?: any): void {
    new YamlFile(project, '.github/workflows/add-to-project.yml', { obj: { ...defaultWorkflow, ...override } });
  }
}
