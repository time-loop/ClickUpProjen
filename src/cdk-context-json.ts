import { awscdk, JsonPatch, SampleFile } from 'projen';

export module cdkContextJson {
  export interface InjectionOptions {
    /**
     * The account number where the role the GH actions will assume lives.
     * This is usually the same account running cdk-pipelines,
     * but can be any account which has been configured with
     *
     * ```ts
     * cdk bootstrap --trust-for-lookup $this_account_number ...
     * ```
     */
    readonly lookupAccountId: string;
    /**
     * Which region should GH Auth into?
     * @default 'us-west-2'
     */
    readonly awsRegion?: string;
    /**
     * How long should the requested role be valid.
     * @default 900 15min seems pretty generous.
     */
    readonly roleDurationSeconds?: number;
  }

  /**
   * Options to support cdk.context.json management
   */
  export interface Options {
    /**
     * @default - do not create stack SampleCode for OIDC role
     */
    readonly createOidcRoleStack?: boolean;
    /**
     * You must configure this if you want to self-mutation cdk.context.json.
     *
     * @default - do not modify the build workflow to authorize against AWS.
     */
    readonly injectionOptions?: InjectionOptions;
  }

  export function injectAwsAuthIntoBuild(project: awscdk.AwsCdkTypeScriptApp, options: InjectionOptions): void {
    const build = project.tryFindObjectFile('.github/workflows/build.yml');
    build?.addOverride('jobs.build.permissions', { 'id-token': 'write' });
    build?.patch(
      JsonPatch.add('/jobs/build/steps/0', {
        name: 'Configure AWS Credentials',
        uses: 'aws-actions/configure-aws-credentials@v2',
        with: {
          'aws-region': options.awsRegion ?? 'us-west-2',
          'role-to-assume': `arn:aws:iam::${options.lookupAccountId}:role/${project.name}-github-actions-role`,
          'role-duration-seconds': options.roleDurationSeconds ?? 900,
        },
      }),
    );
  }

  export function addOidcRoleStack(project: awscdk.AwsCdkTypeScriptApp): void {
    new SampleFile(project, `${project.srcdir}/github-actions-oidc-cdk-context-lookup-role.ts`, {
      contents: `import { core } from '@time-loop/cdk-library';
import { aws_iam, Stage } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Namer } from 'multi-convention-namer';

export class GitHubActionsOIDCCdkContextLookupRole extends core.Stack {
  static asStage(scope: Construct, stageName: string, stageProps: core.StageProps): Stage {
    return new (class extends Stage {
      constructor() {
        super(scope, stageName, stageProps);
        new GitHubActionsOIDCCdkContextLookupRole(this, stageProps);
      }
    })();
  }
  constructor(scope: Construct, props: core.StackProps) {
    const projectName = '${project.name}';
    // NOTE: you should never deploy this anywhere but the cdkPipelines account
    // And the GitHubActionsOIDCPermissions class for cdk-diff should never be deployed
    // in the cdkPipelines account. Name collision is intentional here.
    let id = new Namer([...projectName.split('-'), 'github', 'actions']);
    super(scope, id, props);

    const githubActionsRoleName = id.addSuffix(['role']).kebab;
    const githubActionsRole = new aws_iam.Role(this, githubActionsRoleName, {
      roleName: githubActionsRoleName,
      assumedBy: new aws_iam.FederatedPrincipal(
        \`arn:aws:iam::\${this.account}:oidc-provider/token.actions.githubusercontent.com\`,
        {
          StringEquals: {
            'token.actions.githubusercontent.com:aud': 'sts.amazonaws.com',
          },
          StringLike: {
            'token.actions.githubusercontent.com:sub': \`repo:time-loop/\${projectName}:*\`,
          },
        },
        'sts:AssumeRoleWithWebIdentity',
      ),
    });

    const githubActionsPolicyName = id.addSuffix(['policy']).kebab;
    const githubActionsPolicy = new aws_iam.Policy(this, githubActionsPolicyName, {
      policyName: githubActionsPolicyName,
      statements: [
        new aws_iam.PolicyStatement({
          actions: [ 'sts:AssumeRole' ],
          resources: [ 'arn:aws:iam::*:role/cdk-*lookup-role*' ],
        }),
      ],
    });

    // Attach IAM policy to IAM role
    githubActionsPolicy.attachToRole(githubActionsRole);
  }
}
`,
    });
  }
}
