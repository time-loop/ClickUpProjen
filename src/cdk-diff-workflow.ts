import { typescript, YamlFile } from 'projen';
import { NodePackage } from 'projen/lib/javascript';

export module cdkDiffWorkflow {
  function createCdkDiffWorkflow(options: CDKDiffOptionsConfig) {
    const defaultWorkflow = {
      name: 'cdk-diff',
      on: {
        pull_request: {},
        workflow_dispatch: {},
      },
      permissions: {
        'id-token': 'write',
        contents: 'write',
        'pull-requests': 'write',
      },
      jobs: {
        diff: {
          'runs-on': 'ubuntu-latest',
          env: {
            CI: 'true',
            GITHUB_PR_NUMBER: '${{github.event.pull_request.number}}',
            REPO_NAME: '${{ github.event.repository.name }}',
          },
          steps: [
            {
              name: 'Checkout',
              uses: 'actions/checkout@v3',
              with: {
                ref: '${{ github.event.pull_request.head.ref }}',
                repository: '${{ github.event.pull_request.head.repo.full_name }}',
              },
            },
            {
              name: 'GitHub Packages authorization',
              run: 'cat > .npmrc <<EOF\n//npm.pkg.github.com/:_authToken=${{ secrets.ALL_PACKAGE_READ_TOKEN }}\n@time-loop:registry=https://npm.pkg.github.com/\nEOF',
            },
            {
              name: 'Setup Node.js',
              uses: 'actions/setup-node@v3',
              with: {
                'node-version': '14.x',
              },
            },
            {
              name: 'Install dependencies',
              run: 'yarn install --check-files',
            },
            {
              name: 'ls stacks in each env',
              run: './node_modules/.bin/cdk ls -l -j > cdk-ls.json\nQA_STACKS=$(jq -r \'[.[].id | select(contains("Qa")) ] | join(" ")\' cdk-ls.json)\nSTG_STACKS=$(jq -r \'[.[].id | select(contains("Staging")) ] | join(" ")\' cdk-ls.json)\nPROD_STACKS=$(jq -r \'[.[].id | select(contains("Prod")) ] | join(" ")\' cdk-ls.json)\necho "QA_STACKS=$QA_STACKS" >> $GITHUB_ENV\necho "STG_STACKS=$STG_STACKS" >> $GITHUB_ENV\necho "PROD_STACKS=$PROD_STACKS" >> $GITHUB_ENV\n',
            },
            {
              name: 'configure qa aws credentials',
              uses: 'aws-actions/configure-aws-credentials@v1',
              with: {
                'role-to-assume': options.oidcQaRoleArn,
                'role-duration-seconds': 900,
                'aws-region': 'us-west-2',
              },
            },
            {
              name: 'diff qa',
              run: "set -o pipefail\n./node_modules/.bin/cdk diff $QA_STACKS --progress=events --staging false -e --ignore-errors --version-reporting false &> >(tee cdk-qa.log)\n./node_modules/.bin/ts-node ./node_modules/@time-loop/cdk-log-parser/src/cdkLogParser.ts cdk-qa.log && echo 'QA_HAS_NO_DIFF=true' >> $GITHUB_ENV\n",
            },
            {
              name: 'configure staging aws credentials',
              uses: 'aws-actions/configure-aws-credentials@v1',
              with: {
                'role-to-assume': options.oidcStagingRoleArn,
                'role-duration-seconds': 900,
                'aws-region': 'us-west-2',
              },
            },
            {
              name: 'diff staging',
              run: "set -o pipefail\n./node_modules/.bin/cdk diff $STG_STACKS --progress=events --staging false -e --ignore-errors --version-reporting false &> >(tee cdk-staging.log)\n./node_modules/.bin/ts-node ./node_modules/@time-loop/cdk-log-parser/src/cdkLogParser.ts cdk-staging.log && echo 'STAGING_HAS_NO_DIFF=true' >> $GITHUB_ENV\n",
            },
            {
              name: 'configure prod aws credentials',
              uses: 'aws-actions/configure-aws-credentials@v1',
              with: {
                'role-to-assume': options.oidcProdRoleArn,
                'role-duration-seconds': 900,
                'aws-region': 'us-west-2',
              },
            },
            {
              name: 'diff prod',
              run: "set -o pipefail\n./node_modules/.bin/cdk diff $PROD_STACKS --progress=events --staging false -e --ignore-errors --version-reporting false &> >(tee cdk-prod.log)\n./node_modules/.bin/ts-node ./node_modules/@time-loop/cdk-log-parser/src/cdkLogParser.ts cdk-prod.log && echo 'PROD_HAS_NO_DIFF=true' >> $GITHUB_ENV\n",
            },
            {
              name: 'cdk-notify',
              run: "curl -L \"https://github.com/karlderkaefer/cdk-notifier/releases/download/v2.3.2/cdk-notifier_$(uname)_amd64.gz\" -o cdk-notifier.gz\ngunzip cdk-notifier.gz && chmod +x cdk-notifier && rm -rf cdk-notifier.gz\n./cdk-notifier --log-file ./cdk-qa.log --repo $REPO_NAME --owner $GITHUB_REPOSITORY_OWNER --token ${{ secrets.GITHUB_TOKEN }} --pull-request-id ${{ github.event.pull_request.number }} -t 'Qa Stacks'\n./cdk-notifier --log-file ./cdk-staging.log --repo $REPO_NAME --owner $GITHUB_REPOSITORY_OWNER --token ${{ secrets.GITHUB_TOKEN }} --pull-request-id ${{ github.event.pull_request.number }} -t 'Staging Stacks'\n./cdk-notifier --log-file ./cdk-prod.log --repo $REPO_NAME --owner $GITHUB_REPOSITORY_OWNER --token ${{ secrets.GITHUB_TOKEN }} --pull-request-id ${{ github.event.pull_request.number }} -t 'Production Stacks'\n",
            },
            {
              name: 'Apply qa has no diff label based on diff status',
              if: "env.QA_HAS_NO_DIFF == 'true'",
              uses: 'actions/github-script@v6',
              with: {
                'github-token': '${{ secrets.GITHUB_TOKEN }}',
                script:
                  'github.rest.issues.addLabels({\n  issue_number: context.issue.number,\n  owner: context.repo.owner,\n  repo: context.repo.repo,\n  labels: ["no-qa-changes"]\n})\n',
              },
            },
            {
              name: 'Apply staging has no diff label based on diff status',
              if: "env.STAGING_HAS_NO_DIFF == 'true'",
              uses: 'actions/github-script@v6',
              with: {
                'github-token': '${{ secrets.GITHUB_TOKEN }}',
                script:
                  'github.rest.issues.addLabels({\n  issue_number: context.issue.number,\n  owner: context.repo.owner,\n  repo: context.repo.repo,\n  labels: ["no-staging-changes"]\n})\n',
              },
            },
            {
              name: 'Apply prod has no diff label based on diff status',
              if: "env.PROD_HAS_NO_DIFF == 'true'",
              uses: 'actions/github-script@v6',
              with: {
                'github-token': '${{ secrets.GITHUB_TOKEN }}',
                script:
                  'github.rest.issues.addLabels({\n  issue_number: context.issue.number,\n  owner: context.repo.owner,\n  repo: context.repo.repo,\n  labels: ["no-prod-changes"]\n})\n',
              },
            },
            {
              name: 'Should auto merge',
              id: 'should-auto-merge',
              if: "env.QA_HAS_NO_DIFF == 'true' && env.STAGING_HAS_NO_DIFF == 'true' && env.PROD_HAS_NO_DIFF == 'true' && github.event.pull_request.user.login == 'cu-infra-svc-git'",
              run: 'echo "auto-merge=true" >> $GITHUB_OUTPUT',
            },
            {
              name: 'Enable automerge',
              uses: 'daneden/enable-automerge-action@v1.0.2',
              if: 'steps.should-auto-merge.outputs.auto-merge',
              with: {
                'github-token': '${{ secrets.PROJEN_GITHUB_TOKEN }}',
                'merge-method': 'SQUASH',
                'allowed-author': 'cu-infra-svc-git',
              },
            },
            {
              name: 'Auto approve',
              uses: 'hmarr/auto-approve-action@v2.2.1',
              if: 'steps.should-auto-merge.outputs.auto-merge',
              with: {
                'github-token': '${{ secrets.GITHUB_TOKEN }}',
              },
            },
          ],
        },
      },
    };

    return defaultWorkflow;
  }

  export interface CDKDiffOptionsConfig {
    /**
     * Name of the QA OIDC role name which contains neccesasry IAM policies to run the CDK diff
     *
     * Example arn: `arn:aws:iam::123123123123:role/squad-github-actions-permissions-squad-cdk-github-actions-role`
     */
    readonly oidcQaRoleArn: string;

    /**
     * Name of the Staging OIDC role name which contains neccesasry IAM policies to run the CDK diff
     *
     * Example arn: `arn:aws:iam::123123123123:role/squad-github-actions-permissions-squad-cdk-github-actions-role`
     */
    readonly oidcStagingRoleArn: string;

    /**
     * Name of the Prod OIDC role name which contains neccesasry IAM policies to run the CDK diff
     *
     * Example arn: `arn:aws:iam::123123123123:role/squad-github-actions-permissions-squad-cdk-github-actions-role`
     */
    readonly oidcProdRoleArn: string;
  }

  export function getCDKDiffOptions(options?: CDKDiffOptionsConfig) {
    return options;
  }

  export function addCdkDiffWorkflowYml(
    project: typescript.TypeScriptProject,
    options: CDKDiffOptionsConfig,
    override?: any,
  ): void {
    new YamlFile(project, '.github/workflows/cdk-diff.yml', {
      obj: { ...createCdkDiffWorkflow(options), ...override },
    });
  }

  export function AddCdkLogParserDependency(pkg: NodePackage) {
    pkg.addDevDeps('@time-loop/cdk-log-parser@0.0.0');
  }
}
