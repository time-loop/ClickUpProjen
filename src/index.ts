export * from './clickup-cdk';
export * from './clickup-ts';
export * from './datadog';
export * from './slack-alert';
export * from './renovate-workflow';
export * from './cdk-diff-workflow';
export * from './add-to-project';
