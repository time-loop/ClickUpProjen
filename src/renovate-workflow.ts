import { typescript, YamlFile, RenovatebotOptions } from 'projen';
import merge from 'ts-deepmerge';

export module renovateWorkflow {
  const defaultWorkflow = {
    name: 'upgrade-main',
    on: {
      // Run daily to update the renovate dependency dashboard issue with new updates
      schedule: [
        {
          cron: '0 0 * * MON-FRI',
        },
      ],
      workflow_dispatch: {},
      // Run immediately on any changes to the renovate config or workflow
      push: {
        branches: ['main'],
        paths: ['renovate.json5', '.github/workflows/renovate.yml'],
      },
      // Run immediately when a box to update a dependency is checked on the renovate dashboard issue
      issues: {
        types: ['edited'],
      },
      // or when the rebase now checkbox is ticked on a renovate PR description
      pull_request: {
        types: ['edited'],
      },
    },
    // Prevent renovate running multiple times in parallel due to several subsequent trigger events occurring
    concurrency: {
      group: "${{ github.workflow }}-${{ github.event_name == 'workflow_dispatch' && github.event_name || '' }}",
    },
    jobs: {
      'upgrade-main': {
        'runs-on': 'ubuntu-latest',
        permissions: {
          contents: 'read',
        },
        steps: [
          {
            name: 'Checkout',
            uses: 'actions/checkout@v3',
          },
          {
            name: 'Self-hosted Renovate',
            uses: 'renovatebot/github-action@v34.56.3',
            // Skip running renovate in a loop when renovate updates the dependency dashboard issue and re-triggers this workflow
            if: "(github.event_name != 'issues' && github.event_name != 'pull_request') || github.actor != 'cu-infra-svc-git'",
            with: {
              // projen creates this config for us
              configurationFile: 'renovate.json5',
              // Run projen as the cu-infra-svc-git user
              // We cannot use the default GITHUB_TOKEN to auth as explained here:
              // https://github.com/renovatebot/github-action#token
              token: 'x-access-token:${{ secrets.PROJEN_GITHUB_TOKEN }}',
            },
            env: {
              // Private github packages auth
              RENOVATE_HOST_RULES:
                '[{"matchHost":"https://npm.pkg.github.com/","hostType":"npm","token":"${{ secrets.ALL_PACKAGE_READ_TOKEN }}"}]',
              RENOVATE_NPMRC: '@${{ github.repository_owner }}:registry=https://npm.pkg.github.com',
              // Only update npm dependencies for now
              RENOVATE_ENABLED_MANAGERS: '["npm"]',
              // Tell renovate to run on this repo, without this renovate won't run
              RENOVATE_REPOSITORIES: '["${{ github.repository }}"]',
              // Bypass schedule when the workflow is manually triggered,
              // otherwise the user has to then go and check the box on the dependency dashboard to
              // receive updates and this is kind of confusing
              RENOVATE_FORCE: "${{ github.event_name == 'workflow_dispatch' && '{\"schedule\":null}' || '' }}",
            },
          },
        ],
      },
    },
  };

  export function getRenovateOptions(customOptions: Partial<RenovatebotOptions> = {}) {
    return merge(
      {
        scheduleInterval: ['before 1am on Monday'],
        ignoreProjen: false,
        ignore: [
          // managed by projen
          'node',
        ],
        overrideConfig: {
          rangeStrategy: 'bump',
          extends: ['config:base', 'group:allNonMajor', 'group:recommended', 'group:monorepos'],
        },
      },
      customOptions,
    );
  }

  export function addRenovateWorkflowYml(project: typescript.TypeScriptProject, override?: any): void {
    new YamlFile(project, '.github/workflows/renovate.yml', { obj: { ...defaultWorkflow, ...override } });
  }
}
