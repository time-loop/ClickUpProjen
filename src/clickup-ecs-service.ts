import * as fs from 'fs';
import * as path from 'path';
import { Component, SampleFile } from 'projen';
import merge from 'ts-deepmerge';
import { clickupTs } from './clickup-ts';
import { renovateWorkflow } from './renovate-workflow';
import { ecsServiceCDWorkflow } from './workflows/ecs-service-cd';
import { ecsServiceCIWorkflow } from './workflows/ecs-service-ci';
import { ecsServiceDeployWorkflow } from './workflows/ecs-service-deploy';

export module clickupEcsService {
  export const defaults: Omit<ClickUpTypeScriptEcsServiceProjectOptions, 'name' | 'serviceName'> = {
    ...clickupTs.defaults,
    workflowNodeVersion: '16',
  };

  export interface ClickUpTypeScriptEcsServiceProjectOptions extends clickupTs.ClickUpTypeScriptProjectOptions {
    readonly deployWorkflowOptions?: ecsServiceDeployWorkflow.EcsServiceDeployOptionsConfig;
    readonly serviceName: string;
    readonly serviceHealthcheckPath?: string;
  }

  export class ClickUpTypeScriptEcsServiceProject extends clickupTs.ClickUpTypeScriptProject {
    readonly workflowNodeVersion: string;
    readonly serviceName: string;
    readonly serviceHealthcheckPath: string;

    constructor(options: ClickUpTypeScriptEcsServiceProjectOptions) {
      const serviceHealthcheckPath = options.serviceHealthcheckPath ?? `${options.serviceName}/health`;
      let mergedOptions = merge(
        defaults,
        {
          deps: clickupTs.deps,
          deployWorkflowOptions: {
            deployEnvs: [ecsServiceDeployWorkflow.ClickUpHarnessEnv.QA],
            serviceHealthcheckPath,
          },
        },
        options,
        {
          // Disable projen's built-in docgen class
          docgen: undefined,
          name: clickupTs.normalizeName(options.name),
          renovatebotOptions: renovateWorkflow.getRenovateOptions(options.renovateOptionsConfig),
        },
      );
      super(mergedOptions);
      this.workflowNodeVersion = mergedOptions.workflowNodeVersion!;
      this.serviceName = options.serviceName;
      this.serviceHealthcheckPath = serviceHealthcheckPath;

      new AppSampleCode(this);

      ecsServiceCIWorkflow.createContinuousIntegrationWorkflow(this, {
        ...mergedOptions,
        nodeVersion: this.workflowNodeVersion,
      });

      ecsServiceCDWorkflow.createContinuousDeliveryWorkflow(this, {
        ...mergedOptions,
        ...mergedOptions.deployWorkflowOptions!,
        nodeVersion: this.workflowNodeVersion,
      });
    }
  }

  class AppSampleCode extends Component {
    constructor(project: ClickUpTypeScriptEcsServiceProject) {
      super(project);

      new SampleFile(project, 'Dockerfile', {
        contents: this.renderSample('Dockerfile.sample'),
      });

      new SampleFile(project, 'scripts/docker-healthcheck.sh', {
        contents: this.renderSample('docker-healthcheck.sh.sample', {
          '<%SERVICE_HEALTHCHECK_PATH%>': project.serviceHealthcheckPath,
        }),
      });

      new SampleFile(project, 'scripts/docker-build.sh', {
        contents: this.renderSample('docker-build.sh.sample', {
          '<%SERVICE_NAME%>': project.serviceName,
        }),
      });
    }

    private getSample(fileName: string): string {
      return fs.readFileSync(path.join(__dirname, 'samples', fileName), 'utf-8');
    }

    private renderSample(sampleFileName: string, vars?: { [toReplace: string]: string }): string {
      let contents = this.getSample(sampleFileName);

      if (vars) {
        const toReplace = Object.keys(vars);
        toReplace.map((key) => {
          contents = contents.replace(new RegExp(key, 'g'), vars[key]);
        });
      }
      return contents;
    }
  }
}
