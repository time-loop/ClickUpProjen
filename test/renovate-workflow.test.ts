import { typescript, Testing } from 'projen';

import { renovateWorkflow } from '../src/renovate-workflow';

describe('addRenovateWorkflowYml', () => {
  test('file added', () => {
    const project = new typescript.TypeScriptProject({
      defaultReleaseBranch: 'main',
      name: 'test',
    });
    renovateWorkflow.addRenovateWorkflowYml(project);
    const synth = Testing.synth(project);
    expect(synth['.github/workflows/renovate.yml']).toMatchSnapshot();
  });
  test('override', () => {
    const project = new typescript.TypeScriptProject({
      defaultReleaseBranch: 'main',
      name: 'test',
    });
    renovateWorkflow.addRenovateWorkflowYml(project, { foo: 'bar' });
    const synth = Testing.synth(project);
    expect(synth['.github/workflows/renovate.yml']).toMatchSnapshot();
  });
});

describe('getRenovateOptions', () => {
  test('merges options', () => {
    const options = renovateWorkflow.getRenovateOptions({ overrideConfig: { dryRun: true } });
    // custom option we set
    expect(options.overrideConfig.dryRun).toBe(true);
    // default option that should be deep merged in
    expect(options.overrideConfig.rangeStrategy).toBe('bump');
  });
});
