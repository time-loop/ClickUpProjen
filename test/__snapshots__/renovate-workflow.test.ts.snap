// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addRenovateWorkflowYml file added 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: upgrade-main
on:
  schedule:
    - cron: 0 0 * * MON-FRI
  workflow_dispatch: {}
  push:
    branches:
      - main
    paths:
      - renovate.json5
      - .github/workflows/renovate.yml
  issues:
    types:
      - edited
  pull_request:
    types:
      - edited
concurrency:
  group: \${{ github.workflow }}-\${{ github.event_name == 'workflow_dispatch' && github.event_name || '' }}
jobs:
  upgrade-main:
    runs-on: ubuntu-latest
    permissions:
      contents: read
    steps:
      - name: Checkout
        uses: actions/checkout@v3
      - name: Self-hosted Renovate
        uses: renovatebot/github-action@v39.0.1
        if: (github.event_name != 'issues' && github.event_name != 'pull_request') || github.actor != 'cu-infra-svc-git'
        with:
          configurationFile: renovate.json5
          token: x-access-token:\${{ secrets.RENOVATEBOT_GITHUB_TOKEN }}
        env:
          RENOVATE_HOST_RULES: '[{"matchHost":"https://npm.pkg.github.com/","hostType":"npm","token":"\${{ secrets.ALL_PACKAGE_READ_TOKEN }}"}]'
          RENOVATE_NPMRC: "@\${{ github.repository_owner }}:registry=https://npm.pkg.github.com"
          RENOVATE_ENABLED_MANAGERS: '["npm"]'
          RENOVATE_REPOSITORIES: '["\${{ github.repository }}"]'
          RENOVATE_FORCE: \${{ github.event_name == 'workflow_dispatch' && '{"schedule":null}' || '' }}
"
`;

exports[`addRenovateWorkflowYml override 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: upgrade-main
on:
  schedule:
    - cron: 0 0 * * MON-FRI
  workflow_dispatch: {}
  push:
    branches:
      - main
    paths:
      - renovate.json5
      - .github/workflows/renovate.yml
  issues:
    types:
      - edited
  pull_request:
    types:
      - edited
concurrency:
  group: \${{ github.workflow }}-\${{ github.event_name == 'workflow_dispatch' && github.event_name || '' }}
jobs:
  upgrade-main:
    runs-on: ubuntu-latest
    permissions:
      contents: read
    steps:
      - name: Checkout
        uses: actions/checkout@v3
      - name: Self-hosted Renovate
        uses: renovatebot/github-action@v39.0.1
        if: (github.event_name != 'issues' && github.event_name != 'pull_request') || github.actor != 'cu-infra-svc-git'
        with:
          configurationFile: renovate.json5
          token: x-access-token:\${{ secrets.RENOVATEBOT_GITHUB_TOKEN }}
        env:
          RENOVATE_HOST_RULES: '[{"matchHost":"https://npm.pkg.github.com/","hostType":"npm","token":"\${{ secrets.ALL_PACKAGE_READ_TOKEN }}"}]'
          RENOVATE_NPMRC: "@\${{ github.repository_owner }}:registry=https://npm.pkg.github.com"
          RENOVATE_ENABLED_MANAGERS: '["npm"]'
          RENOVATE_REPOSITORIES: '["\${{ github.repository }}"]'
          RENOVATE_FORCE: \${{ github.event_name == 'workflow_dispatch' && '{"schedule":null}' || '' }}
foo: bar
"
`;

exports[`getRenovateOptions defaults 1`] = `
{
  "ignore": [
    "node",
    "@time-loop/clickup-projen",
  ],
  "ignoreProjen": true,
  "labels": [
    "renovate",
  ],
  "overrideConfig": {
    "automergeType": "pr",
    "extends": [
      "config:base",
      "group:recommended",
      "group:monorepos",
      "github>whitesource/merge-confidence:beta",
    ],
    "packageRules": [
      {
        "addLabels": [
          undefined,
        ],
        "allowedVersions": "!/^[0-9]+\\.[0-9]+\\.[0-9]+-(alpha|beta).*$/",
        "automerge": undefined,
        "groupName": "all non-major dependencies",
        "groupSlug": "all-minor-patch",
        "matchPackagePatterns": [
          "*",
        ],
        "matchUpdateTypes": [
          "minor",
          "patch",
        ],
      },
      {
        "addLabels": [
          "optional",
        ],
        "matchDepTypes": [
          "optionalDependencies",
        ],
      },
    ],
    "platformAutomerge": true,
    "prConcurrentLimit": 0,
    "prHourlyLimit": 0,
    "rangeStrategy": "bump",
  },
  "scheduleInterval": [
    "before 1am on Monday",
  ],
}
`;
